### ISSUE TITLE
Rule proposal: prefer-string-function-replacer

### ISSUE THREAD

**Issue by xiaoxiangmoe** [typescript-eslint/typescript-eslint#10575]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Prefer using function replacer than string replacer in String#replace and String#replaceAll.

for example:

```js
'hello {} world'.replace('{}', () => "A$'1")

"hello A$'1 world" // This is good
```

```js
'hello {} world'.replace('{}', "A$'1")

'hello A world1 world' // This is too bad
```

### Fail Cases

```TypeScript
'hello {} world'.replace('{}', "A$'1")

function test(str: string, replacer: string) {
  return str.replaceAll('{}', replacer)
}
```

### Pass Cases

```TypeScript
'hello {} world'.replace('{}', () => "A$'1")

function test(str: string, replacer: () => string) {
  return str.replaceAll('{}', replacer)
}

function test(str: string, replacer: string | (() => string)) {
  return str.replace('{}', replacer)
}

function test(str: string, replacer: any) {
  return str.replace('{}', replacer)
}
```

**Comment by bradzacher** [comment]
I don't think this is a strict, non-opinionated enough proposal for us to consider for this project.
It requires types for correctness, yes, but it's not typescript specific.
The checks it provides are stylistic - many (most) people wouldn't want to always use the function form and would be okay using both in their codebase. We're not looking to add stylistic rules unless there's a really good reason to prefer one style over the other.

For the above reason I'm in favour of rejecting this proposal. Cc @typescript-eslint/triage-team

**Comment by Josh-Cena** [comment]
Yes. I don't believe the chance of false positives is high enough to prevent it from being included in a type-unaware plugin like eslint-plugin-unicorn. That said, I do see why you want to do that: I myself have been bitten by the string form! https://x.com/SidaChen63/status/1722672650403614781

**Comment by JoshuaKGoldberg** [comment]
Yeah, I'm -1 on this as well. In a perfect world it might be nice to get the safety proposed in this rule. But we also have to balance rules with how likely users are to be annoyed by them. We've found in the past that if we provide rules which add bloat to many uses for the sake of fixing rare edge cases, we end up turning users away from linting altogether. 😞 

Closing this issue out for typescript-eslint core, but it's a good idea for a rule! https://typescript-eslint.io/developers has docs on how you could make this a standalone rule / plugin.
